haas/ext/switches/dell.py

```diff
@@ -83,60 +113,129 @@ def connect(switch):
         self.disable_vlan(vlan_id)
-        self._sendline('sw trunk native vlan none')
-
-    def disconnect(self):
-        if self._should_save('dell'):
-            self._save_running_config()
+        # first set the dummy vlan as trunking vlan, then set that as it's
+        # native, then remove that vlan from trunking vlans. otherwise the
+        # switch won't let you set a native vlan that isn't added.
+        self._sendline('sw trunk allowed vlan add ' + self.dummy_vlan)
+        self._sendline('sw trunk native vlan ' + self.dummy_vlan)
+        self._sendline('sw trunk allowed vlan remove ' + self.dummy_vlan)
+
+    def _set_native(self, old_native, network_id, interface):
+        # create the vlan that we need to set as native
+        self._sendline('exit')
+        self._sendline('vlan ' + network_id)
         self._sendline('exit')
-        self.console.expect(pexpect.EOF)
-        logger.debug('Logged out of switch %r', self.switch)
+        self._sendline('exit')
+        self.enter_if_prompt(interface)
+        # set the native vlan here
+        self.set_native(old_native, network_id)
+
+    def _int_config(self, interface):
+        """Collect information about the specified interface
+
+        Returns a dictionary from the output of ``show int sw <interfaces>``.
+        """
+
+        self._sendline('show int sw %s' % interface)
+        self.console.expect('Port: .*')
+        k, v = 'key', 'value'
+        result = {k: v}
+        key_lines = self.console.after.splitlines()
+        del key_lines[-3:]
+        for line in key_lines:
+            k, v = line.split(':', 1)
+            result[k] = v
+        # expecting main_prompt here fails, because it appears that the
+        # main_prompt is a part of the interface configuration (console.after)
+        # sending a new line clears things up here.
+        self._sendline('\n')
+        self.console.expect(self.main_prompt)
+        return result
 
     def get_port_networks(self, ports):
         num_re = re.compile(r'(\d+)')
         port_configs = self._port_configs(ports)
         result = {}
         for k, v in port_configs.iteritems():
-            native = v['Trunking Native Mode VLAN'].strip()
+            native = v['Trunking Mode Native VLAN'].strip()
             match = re.match(num_re, native)
             if match:
                 # We need to call groups to get the part of the string that
                 # actually matched, because it could include some junk on the
                 # end, e.g. "100 (Inactive)".
                 num_str = match.groups()[0]
                 native = int(num_str)
+                if native == int(self.switch.dummy_vlan):
+                    native = None
             else:
                 native = None
             networks = []
-            range_str = v['Trunking VLANs Enabled']
-            for range_str in v['Trunking VLANs Enabled'].split(','):
+            range_str = v['Trunking Mode VLANs Enabled']
```

Comment:
I realize this was already there, but -- `range_str` is reassigned on the next line, so this doesn't do anything.